Refactor(pyavd): Refactor backend ANTA code of PyAVD (#6262)
aristanetworks/ansible-avd: docs/porting-guides/6.x.x.md

) -> AntaCatalog
```

> **Action Required:** Update your code to convert structured configurations into `AvdFabricData` before calling the function.

Reviewer: we may need to sync on this so that we are consistent in the portigng guide entries for these signature changes. Lets keep track in the main umbrella issue

Author: Agreed. As discussed live we will sync during documentation week :)

## Changes to PyAVD

### Changes to `pyavd.get_device_test_catalog`

In AVD 6.0.0, `pyavd.get_device_test_catalog` has undergone a few changes to improve maintainability.

#### 1. Updated Function Signature

The function now accepts an `AvdFabricData` dataclass instance instead of a dictionary for the `minimal_structured_configs` argument. The argument name has changed to `fabric_data` for clarity.

```diff

>   def get_device_test_catalog(

Reviewer 2: is this space change expected?

Author: Yes.

+     structured_config=structured_configs["DC1-LEAF1A"],
+     fabric_data=fabric_data
+ )
```

#### 2. Removal of `ignored_is_deployed` Setting

The `AvdCatalogGenerationSettings.ignored_is_deployed` attribute has been removed.

- **AVD 5.x:** The function checked `metadata.is_deployed` inside the structured config. If `False`, it returned an empty catalog *unless* `ignored_is_deployed` was set to `True`.
- **AVD 6.0.0:** The function *always* generates a catalog, ignoring the `metadata.is_deployed` key.

**Action Required:** The caller is now responsible for checking the `is_deployed` status before requesting a catalog.

```diff
  from pyavd import get_device_test_catalog
  from pyavd.api._anta import AvdFabricData

  dc1_leaf1a_structured_config = {"hostname": "DC1-LEAF1A", "metadata": {"is_deployed": False}}

  structured_configs = {"DC1-LEAF1A": dc1_leaf1a_structured_config}
  fabric_data = AvdFabricData.from_structured_configs(structured_configs)

- anta_catalog = get_device_test_catalog(hostname="DC1-LEAF1A", structured_config=dc1_leaf1a_structured_config, fabric_data=fabric_data)
+ if dc1_leaf1a_structured_config.get("metadata", {}).get("is_deployed", True) is False:

> +     anta_catalog = AntaCatalog()

Reviewer: do we want to suggest generating empty catalogs?

Author: Nop. Fixed in https://github.com/aristanetworks/avd/pull/6262/commits/002d25436dae9f07cda3d83fd701abb6ce9fed61.

+ else:
+     anta_catalog = get_device_test_catalog(hostname="DC1-LEAF1A", structured_config=dc1_leaf1a_structured_config, fabric_data=fabric_data)
```

#### 3. Removal of `input_factory_settings` Setting

The `AvdCatalogGenerationSettings.input_factory_settings` attribute, previously used to globally enable BGP validation for VRF peers, has been removed. Control has moved to the device structured configuration.

> **Action Required:** Instead of passing `InputFactorySettings(allow_bgp_vrfs=True)`, you must now set `metadata.validate_state` to `True` on specific BGP neighbors or neighbor interfaces (RFC 5549) within the structured configuration.

Reviewer: We had a debate with Claus this morning on another PR and decided to remove the space so editing for alignment

```suggestion
**Action Required:** Instead of passing `InputFactorySettings(allow_bgp_vrfs=True)`, you must now set `metadata.validate_state` to `True` on specific BGP neighbors or neighbor interfaces (RFC5549) within the structured configuration.
```

Author: All good. Fixed in https://github.com/aristanetworks/avd/pull/6262/commits/002d25436dae9f07cda3d83fd701abb6ce9fed61.

```diff
  from pyavd import get_device_test_catalog
- from pyavd.api._anta import AvdCatalogGenerationSettings, AvdFabricData, InputFactorySettings
+ from pyavd.api._anta import AvdCatalogGenerationSettings, AvdFabricData

  # Update structured config to enable validation per neighbor/VRF
- dc1_leaf1a_structured_config = {"hostname": "DC1-LEAF1A"}
+ dc1_leaf1a_structured_config = {
+     "hostname": "DC1-LEAF1A",
+     "router_bgp": {
+         "vrfs": [{
+             "name": "PROD",
+             "neighbors": [{
+                 "ip_address": "10.1.1.11",
+                 "metadata": {"validate_state": True}
+             }]
+         }]
+     }
+ }

  structured_configs = {"DC1-LEAF1A": dc1_leaf1a_structured_config}
  fabric_data = AvdFabricData.from_structured_configs(structured_configs)

- input_factory_settings = InputFactorySettings(allow_bgp_vrfs=True)
- settings = AvdCatalogGenerationSettings(input_factory_settings=input_factory_settings)
- anta_catalog = get_device_test_catalog(hostname="DC1-LEAF1A", structured_config=dc1_leaf1a_structured_config, fabric_data=fabric_data, settings=settings)

> + anta_catalog = get_device_test_catalog(hostname="DC1-LEAF1A", structured_config=dc1_leaf1a_structured_config, fabric_data=fabric_data)

Reviewer: This may be clearer than the long list
```suggestion
  anta_catalog = get_device_test_catalog(
      hostname="DC1-LEAF1A",
      structured_config=dc1_leaf1a_structured_config,
      fabric_data=fabric_data,
-     settings=settings
  )
```

Author: Nice. I adopted this style in other places as well in https://github.com/aristanetworks/avd/pull/6262/commits/002d25436dae9f07cda3d83fd701abb6ce9fed61.

+ from pyavd._anta.constants import StructuredConfigKey
+ from pyavd._anta.input_factories._decorators import skip_if_missing_config
- from pyavd.api._anta import AvdCatalogGenerationSettings, TestSpec
+ from pyavd.api._anta import AvdCatalogGenerationSettings, AvdTestSpec

  class CustomVerifyReachabilityInputFactory(AntaTestInputFactory[VerifyReachability.Input]):
      """Custom input factory class for the `VerifyReachability` test."""

+     @skip_if_missing_config(StructuredConfigKey.ETHERNET_INTERFACES)
      def create(self) -> Iterator[VerifyReachability.Input]:
          # Custom logic to generate the inputs
          ...

- test_spec = TestSpec(
-     test_class=VerifyReachability,
-     conditional_keys=[StructuredConfigKey.ETHERNET_INTERFACES],
-     input_factory=CustomVerifyReachabilityInputFactory
- )
+ test_spec = AvdTestSpec(
+     test_class=VerifyReachability,
+     input_factory=CustomVerifyReachabilityInputFactory
+ )

  settings = AvdCatalogGenerationSettings(custom_test_specs=(test_spec,))
```

#### 5. `AvdCatalogGenerationSettings` is now a Dataclass

`AvdCatalogGenerationSettings` has been converted from a Pydantic model to a standard Python `dataclass`.

> **Action Required:** If you code relies on Pydantic-specific methods (like validation error handling or serialization), updates are required.

Reviewer: ```suggestion
**Action Required:** If your code relies on Pydantic-specific methods (like validation error handling or serialization), updates are required.
```

Author: Fixed in https://github.com/aristanetworks/avd/pull/6262/commits/002d25436dae9f07cda3d83fd701abb6ce9fed61.
